File: hello_ws/src/hello_action_test_cpp/src/hello_server.cpp
// Server side (using Action)
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp" // may not be required if implemented into rclcpp
#include "hello_msgs/action/hello.hpp"

class HelloServer : public rclcpp::Node
{
public:
	/* call using HelloServer::hoge when out of scope */
    using Hello = hello_msgs::action::Hello;
    using GoalHandleHello = rclcpp_action::ServerGoalHandle<Hello>;

    explicit HelloServer(const rclcpp::NodeOptions &options);
private: //variables and instances
    rclcpp_action::Server<Hello>::SharedPtr action_server;
private: //functions
    rclcpp_action::GoalResponse handle_goal(
        const rclcpp_action::GoalUUID &uuid,
        std::shared_ptr<const Hello::Goal> goal
    );
    rclcpp_action::CancelResponse handle_cancel(const std::shared_ptr<GoalHandleHello> goal_handle);
    void execute(const std::shared_ptr<GoalHandleHello> goal_handle);
    void handle_accepted(const std::shared_ptr<GoalHandleHello> goal_handle);
};
HelloServer::HelloServer(const rclcpp::NodeOptions &options=rclcpp::NodeOptions())
    : Node("hello_action_test_server", options)
{
    using namespace std::placeholders;

    this->action_server = rclcpp_action::create_server<Hello>(
        this->get_node_base_interface(),
        this->get_node_clock_interface(),
        this->get_node_logging_interface(),
        this->get_node_waitables_interface(),
        "hello_action",
        std::bind(&HelloServer::handle_goal, this, _1, _2),
        std::bind(&HelloServer::handle_cancel, this, _1),
        std::bind(&HelloServer::handle_accepted, this, _1)
    );
}
rclcpp_action::GoalResponse HelloServer::handle_goal(
	const rclcpp_action::GoalUUID &uuid,
	std::shared_ptr<const HelloServer::Hello::Goal> goal
)
{
	RCLCPP_INFO(this->get_logger(), "Received goal request:\n\thello_goal: %s", goal->hello_goal.c_str());
	(void)uuid; //why???
	if (goal->hello_goal != "Hello World!") //reject goal that is not "Hello World!"
		return rclcpp_action::GoalResponse::REJECT;
	return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
}
rclcpp_action::CancelResponse HelloServer::handle_cancel(const std::shared_ptr<HelloServer::GoalHandleHello> goal_handle)
{
	RCLCPP_INFO(this->get_logger(), "Received request to cancel goal.");
	(void)goal_handle; //why???
	return rclcpp_action::CancelResponse::ACCEPT;
}
void HelloServer::execute(const std::shared_ptr<HelloServer::GoalHandleHello> goal_handle)
{
	RCLCPP_INFO(this->get_logger(), "Executing goal.");
	rclcpp::Rate loop_rate(1);
	const auto goal = goal_handle->get_goal(); // instance to store goal
	auto feedback = std::make_shared<Hello::Feedback>(); // making instance for feedback
	/* initializing feedback */
	std::string &status = feedback->status;
	status = "";

	auto result = std::make_shared<Hello::Result>(); // making instance for result

	bool complete = false;

	while (!complete)
	{
		if (goal_handle->is_canceling())
		{
			result->status = status;
			goal_handle->canceled(result);
			RCLCPP_INFO(this->get_logger(), "Goal Canceled.");
			return ;
		}
		status += goal->hello_goal[status.length()]; // update Hello status

		/* Publish Feedback */
		RCLCPP_INFO(this->get_logger(), "Publishing feedback: %s", status.c_str());
		goal_handle->publish_feedback(feedback);

		complete = status == goal->hello_goal;

		loop_rate.sleep();
	}

	if (rclcpp::ok())
	{
		result->status = status;
		goal_handle->succeed(result);
		RCLCPP_INFO(this->get_logger(), "Goal Succeeded!");
	}
}
void HelloServer::handle_accepted(const std::shared_ptr<HelloServer::GoalHandleHello> goal_handle)
{
	using namespace std::placeholders;
	
	std::thread{std::bind(&HelloServer::execute, this, _1), goal_handle}.detach();
}

int main(int argc, char **argv)
{
	rclcpp::init(argc, argv);
	
	auto hello_action_server = std::make_shared<HelloServer>();

	rclcpp::spin(hello_action_server);
	rclcpp::shutdown();
}

File: hello_ws/src/hello_action_test_py3/hello_action_test_py3/hello_server.py
# hello_server
import rclpy
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from rclpy.callback_groups import ReentrantCallbackGroup
from rclpy.executors import MultiThreadedExecutor # may not be required
from rclpy.node import Node

from hello_msgs.action import Hello

# The Server handles the main process while the Client just orders the Server what it wants
class HelloServer(Node):
    def __init__(self):
        super().__init__('hello_action_test_server')

        self.hello_server = ActionServer(self, Hello, 'hello_action',
                execute_callback=self.executeCB,
                callback_group=ReentrantCallbackGroup(),
                goal_callback=self.goalCB,
                cancel_callback=self.cancelCB)

    def destroy(self): # cleanup
        self.hello_server.destroy()
        self.destroy_node()

    def goalCB(self, goal_request): # callback for goal
        ### Accepts or rejects a client request to begin an action. ###
        self.get_logger().info('Received goal request:\n{}'.format(goal_request))
        return GoalResponse.ACCEPT # returns to client that a request to begin an action has been accepted.

    def cancelCB(self, goal_handle):
        ### Accepts or rejects a client request to cancel an action. ###
        self.get_logger().info('Received cancel request:\n{}'.format(goal_handle))
        return CancelResponse.ACCEPT # returns to client that a request to cancel an action has been accepted.

    async def executeCB(self, goal_handle): # corutine for executing goal
        self.get_logger().info('Executing goal...')

        # Initial input for Hello
        feedback_msg = Hello.Feedback()
        feedback_msg.status = ''

        complete = False

        # Start execution of action
        while not complete:
            if goal_handle.is_cancel_requested:
                goal_handle.canceled()
                self.get_logger().info('Goal canceled')
                return Hello.Result()

            # update Hello status
            feedback_msg.status += goal_handle.request.hello_goal[len(feedback_msg.status)]
            
            # publish feedback
            self.get_logger().info('Publishing feedback: {0.status}'.format(feedback_msg))
            goal_handle.publish_feedback(feedback_msg)

            complete = feedback_msg.status == goal_handle.request.hello_goal

        goal_handle.succeed()

        # Populate result message
        result = Hello.Result()
        result.status = feedback_msg.status

        self.get_logger().info('Result: {0.status}'.format(result))

        return result

def main(args=None):
    rclpy.init(args=args)

    hello_server = HelloServer()

    executor = MultiThreadedExecutor()

    try:
        rclpy.spin(hello_server, executor=executor)
    except KeyboardInterrupt:
        pass
    hello_server.destroy()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
